### DISCUSSION TITLE
VSCode - tsconfig.dev.json file not used for development

### DISCUSSION THREAD

**Discussion by dkershner6** [projen/projen#3202] [Q&A] [answered]
Hey projen community, really appreciate this package.

I feel a little crazy for asking this, as I have seen very little regarding it from others, but projen seems heavily established around using `tsconfig.dev.json` for development (and the normal `tsconfig.json` for publishing/releasing).  VSCode, on the other hand, only allows the `tsconfig.json` filename for development. [2016 Issue](https://github.com/microsoft/vscode/issues/12463).

For non-JSII projects, I can just change the names and it is no biggie, but doing so feels against the grain, and doesn't help me on JSII projects.  Is there another way to handle this?  A setting I am missing? How does everyone else achieve proper type-checking?

At the moment, I have to write different code patterns in test files than production files to pass type-checking in VSCode, and more importantly, the type-checking in VSCode doesn't match the actuality.

Happy to close this if I am missing something, but at least it will be here for someone equivalently confused as I.

**Comment by mrgrain** [comment]
I think you are right, the number one reason for this is jsii. Because jsii requires some annoyingly strict TS config, and most people (Probably? Who knows 🤷‍♂️) like to write that cases a bit less strict.

Now by default, the main difference between the configs should be the included files. 

I'm not sure what issue you run in exactly, maybe you could share a few concrete examples?


VSCode not only defaults to `tsconfig.json` - it also uses its own config. In particular included files, because it will analyze every file that you open.

PS: Converted this to a discussion for now. Feels more appropriate maybe?! If this distills down to a more concrete feature request or bug report we can open a new one.

**Reply by dkershner6** [reply]
> I'm not sure what issue you run in exactly, maybe you could share a few concrete examples?

Of course! Something to keep in mind is that I am assessing projen to convert our entire company's repositories to, as a way to more easily share config, so I will be using this in a very non-jsii fashion (more typical Typescript software dev).

All below is true for things that are included in the dev file, but not in the publish file (or VSCode default, for JSII, until you run a build, then the generated JSII file).

Basics:
- Intellisense does not work, i.e. VSCode will not help you find local names, code, etc.  Neither does any other IDE related feature for typescript.

> more importantly, the type-checking in VSCode doesn't match the actuality.

To expand on this, any differences between the configs will spark a conflict, as an example:

`lodash` and its children are exported using `=`.

```typescript
import { set } from "lodash";
export = set;
```

For a file within src, if you write

```typescript
import set from "lodash.set";
```

You receive a type-error: `Module '"some-path/node_modules/@types/lodash.set/index"' can only be default-imported using the 'esModuleInterop' flag ts(1259)`

Changing this to

```typescript
import set = require("lodash.set");
```

Fixes this, builds with no issue.

Then I (or you) go to test this same component.  In your test file, you write:

```typescript
import set = require("lodash.set");
```

You now receive two errors, one from TS, one from ESLint:

```
Import assignment cannot be used when targeting ECMAScript modules. Consider using 'import * as ns from "mod"', 'import {a} from "mod"', 'import d from "mod"', or another module format instead.ts(1202)
A `require()` style import is forbidden.eslint[@typescript-eslint/no-require-imports](https://typescript-eslint.io/rules/no-require-imports)
```

Changing to a standard default import fixes this issue.

This is just an example, but basically applies to any rule that differs between the configs.

**Reply by mrgrain** [reply]
Gotcha. For clarification: Is jsii in the mix because you are looking at using projen with a Construct library? And you are going to use the `AwsCdkConstructLibrary`project type?

**Reply by dkershner6** [reply]
To be more clear on our intent:

_Sometimes_ JSII is in the mix, most of the time it isn't.  As in, we do have custom AWS CDK Constructs, where we use `AwsCdkConstructLibrary`, but most of the time we are just using `TypescriptProject`, maybe a `AwsCdkTypeScriptApp` or just a `TypescriptAppProject`.

On all of the latter, I can just pull a switcheroo and it seems fine, but I cannot fix JSII.

A switcheroo:

```typescript
export const RECOMMENDED_TSCONFIG_NODE_20: Partial<TypeScriptProjectOptions> = {
    tsconfig: {
        fileName: "tsconfig.publish.json",
        compilerOptions: {},
    },
    tsconfigDev: {
        fileName: "tsconfig.json", // Shim for VSCode, dev file must be named this
        compilerOptions: {},
    },
};
```

This works, but _feels_ wrong, and I doubt I am alone here.

Basically I want to expand my personal OSS and my company's `projen` use _beyond_ JSII to _everything_, but I need to ask if this is a fool's errand, because it is not meant to be used that way, before I jump in with both feet.

In addition, even with JSII, the dev experience is less than ideal, at least in VSCode.

**Reply by mrgrain** [reply, marked as answer]
Okay, I think I've got you know. There is a lot to unpack here.

### A general remark to begin with

(I will use "opinion" as a general term referring vaguely to how a certain feature of any of the software involved might be used.)

You are (looking at) migrating an existing project with opinions to use a new tool that comes with lots of opinions (at least by default). This is always going to be a challenge in one way or another. For each opinion you will have to decide if you change the way you currently work and use "the projen way" or if you want to (try to) change how projen does things by default. Neither is better or worse. Depending on the specific opinion it might be easier to go with one of the approaches. It is also worth noting the distinction between: **projen the framework**, which will let you do pretty much anything you want in exactly the way you want it - it might just take time & effort; and **projen the batteries-included project types**, which might not or at least you have to "reverse engineer" the opinions you don't want.

---

### Part 1: Let's ignore jsii for a bit

> Intellisense does not work, i.e. VSCode will not help you find local names, code, etc. Neither does any other IDE related feature for typescript.

For me VSCode does seems to be making some reasonable default assumptions if no config is found. I get all of the features you describe as missing out of the box. That kind of makes sense since VSCode explicitly wants works for vanilla JavaScript as well. I'm not sure why your experience is much different. However this shouldn't matter to much, because a config can always be present.

> but projen seems heavily established around using tsconfig.dev.json for development (and the normal tsconfig.json for publishing/releasing).

That's right, but by default they are functionally the same. The difference is just the files that are included/excluded and the target outdir. This is mostly so that test files don't get compiled but run with `ts-jest` instead (though you can also force compiled tests).

> For non-JSII projects, I can just change the names and it is no biggie

You can, if you'd like to. But you don't have to. You can set `disableTsconfigDev: true` and/or follow [the examples in the migration guide](https://projen.io/typescript.html#default-directory-structure) to adjust your config even further. Now you unfortunately cannot use that setting for jsii. See below for a further discussion.


> but doing so feels against the grain
> This works, but _feels_ wrong

It shouldn't. If there's an option it is supposed to be used. You might just feel that your (company's) opinions don't match with the ones that projen ships with. And that's fine. You will probably end up building a set of customized project types for your company that come with opinions that you like better. [I'm just a dude and even I do this for my OSS projects](https://github.com/mrgrain/projen-projects).

> and I doubt I am alone here.

IDK, but this hasn't been raised a lot yet. My guess is that somehow for most people VSCode's default config is good enough. And if not running the tests will catch it. I'm mildly surprised this hasn't come up yet.

The name switcheroo you are doing is a great idea. It might just not needed for most cases. But hey -  I definitely wasn't aware of this in that much detailed yet. I'm glad you've found it.

--- 

### Part 2: Re jsii

I have some bad news for you: I don't think there is a good solution to this. Just a number of different trade-offs you can make.

Here are the limitations I take into account:

- jsii MUST generate the `tsconfig.json` file
- jsii CANNOT use a different file name
- VSCode ONLY uses `tsconfig.json`

This really just leaves you with two bad options:

- Accept that the Intellisense for your test files is incorrect (will use VSCode defaults instead)
- Include test files in the jsii tsconfig, and subject yourself to (unnecessary) strict rules

projen currently assumes that the former is the less worse of the options.

> Intellisense does not work, i.e. VSCode will not help you find local names, code, etc. Neither does any other IDE related feature for typescript.

I'm guessing you are referring to the fact that `JsiiProjects` ignore the `tsconfig.json` file from git by default and/or that the file is not created by projen. That's because jsii must always manage the file and jsii recommends to not include it into git. The latter you can easily fix:

```
project.gitignore.removePatterns('tsconfig.json');
```

> You can set `disableTsconfigDev: true`

This is a quote from me from above. Unfortunately you currently can't set this for `JsiiProjects`. The reason you can't is a bit silly and I might be inclined to consider this a bug. Although I don't know what a good fix is, accept for lowering the validation from an error to a warning.

> In addition, even with JSII, the dev experience is less than ideal, at least in VSCode.

The dev experience with jsii is less than ideal. Full stop. 😂

**Reply by dkershner6** [reply]
> It is also worth noting the distinction between: projen the framework, which will let you do pretty much anything you want in exactly the way you want it - it might just take time & effort; and projen the batteries-included project types, which might not or at least you have to "reverse engineer" the opinions you don't want.

Just want to say that I have a firm appreciation for this, I just had a 100% clear opinion from VSCode, and a very clear opinion from `projen` in the opposite direction, so figured it was worth asking.  I haven't asked about all of the other places where my (our) opinions differ from projen's. 😆 

> For me VSCode does seems to be making some reasonable default assumptions if no config is found. I get all of the features you describe as missing out of the box. That kind of makes sense since VSCode explicitly wants works for vanilla JavaScript as well. I'm not sure why your experience is much different. However this shouldn't matter to much, because a config can always be present.

I am going to just concede this one.  I have definitely experienced oddities on "non-included" files way more than "included" ones, but you are correct, it does mostly work, especially if you have been working in the area you are attempting to "intellisense".  I tried to find a reliable replication, and VSCode is not cooperating with me (too much cooperating?).

> The dev experience with jsii is less than ideal. Full stop. 😂

Hey, I was being nice! ;)

**Perhaps what I mean is a different question then: Should defaults for non-jsii projects be switcheroo'd?**

Another question you can feel free not to answer, but seems worth asking since you would probably know.: Can the jsii build be somehow nested in a folder and THAT folder has the `tsconfig.json` file in it (allowing for the root to be normal?).

In any case, I appreciate the discourse, it gives me much more solid footing to make decisions.  Thank you.

**Reply by mrgrain** [reply]
> Perhaps what I mean is a different question then: Should defaults for non-jsii projects be switcheroo'd?

I supposed that would be a reasonable possibility. Not sure how to make this decision though (as much as I love VSCode, other IDEs exists) and I worry it will be a huge pain to manage the migration for existing projects.

> Another question you can feel free not to answer, but seems worth asking since you would probably know.: Can the jsii build be somehow nested in a folder and THAT folder has the tsconfig.json file in it (allowing for the root to be normal?).

In principle I can't see why not. projen is definitely not setup for this right now. As a work around you could build some sort of mini-monorepo with a `JsiiProject` in a subdirectory and a `TypeScriptProject` that contains the test cases and imports the jsii projects. That might not be too hard to make it work.

**Reply by mrgrain** [reply]
> Just want to say that I have a firm appreciation for this, I just had a 100% clear opinion from VSCode, and a very clear opinion from projen in the opposite direction, so figured it was worth asking. I haven't asked about all of the other places where my (our) opinions differ from projen's. 😆

Along the note of my other comment ("as much as I love VSCode, other IDEs exists"), I think you might find that some of the original creators of projen did not use VSCode. 😁

**Reply by dkershner6** [reply]
I have what I need to make decisions, and i'll leave it here with your answer marked.  Thank you again for taking the time.

> as much as I love VSCode, other IDEs exists

Oh yes, but I am guessing that other IDEs would accept a `tsconfig.json` file as the one to use. ;) (But migrating existing heard).  I was pretty disappointed in our beloved VSCode for this particular non-setting, to be sure.

**Reply by dkershner6** [reply]
@mrgrain , as I was doing one of your PR comments, I finally have a solid reason to have these files working for VSCode:

The ultra useful `Rename Symbol` capability (rename anything, but only the properly referenced version), does not function in files outside your tsconfig.json `includes`.  i.e. Any reference in a test file to the thing you are renaming would not be renamed.

I am quite sure there are plenty of other things that are somewhat like this that behave similarly, I don't do a ton of JSII development, and have this problem fixed elsewhere, but ill keep my eyes open when I do.

**Reply by mrgrain** [reply]
And I'm sure there a more! I remembered a reason why the prod `tsconfig.json` should be used: It might be stricter than the dev version. Mostly because one might allow things in tests that are not allowed in prod. Not sure if that's a good enough reason. I wonder if longterm we can default to a single config (at least for non jsii).

**Reply by dkershner6** [reply]
> And I'm sure there a more! I remembered a reason why the prod `tsconfig.json` should be used: It might be stricter than the dev version. Mostly because one might allow things in tests that are not allowed in prod. Not sure if that's a good enough reason. I wonder if longterm we can default to a single config (at least for non jsii).

We agree that prod shouldn't be used as the primary tsconfig for the weird handling of tests, I went into great detail here on some wacky behavior: https://github.com/projen/projen/discussions/3202#discussioncomment-7958381
